File: solutions/1723/main.py
from typing import * 

# TLE，为啥？？
class Solution:
    def minimumTimeRequired(self, jobs: List[int], k: int) -> int:
        state2time = lambda state: sum(jobs[i] for i, j in enumerate(jobs) if state & (1 << i)) 
        # init dp table 
        n = len(jobs)
        dp = [[0] * (2 ** n) for _ in range(k)]
        for state in range(2 ** n): 
            dp[0][state] = state2time(state) 
        # traverse 
        for i in range(1, k):
            for state in range(2**n): 
                substate = state 
                min_c = float('inf') 
                while substate > 0:  
                    min_c = min(min_c, max(dp[i-1][state^substate], dp[0][substate]))
                    substate = (substate - 1) & state 
                dp[i][state] = min_c 
        return dp[-1][-1] 


if __name__ == "__main__": 
    s = Solution() 

    jobs = [3,2,3]
    k = 3
    assert s.minimumTimeRequired(jobs, k) == 3 

    jobs = [20010,20006,20014,20004,20008,20006,20005,20012,19999,20014,20003,20012]
    k = 8
    assert s.minimumTimeRequired(jobs, k) == 40011

File: solutions/1723/main.rs
impl Solution {
    pub fn state_to_time(jobs: &[i32], state: i32) -> i32 {
        let mut state = state; 
        let mut time = 0;
        let mut count = 0; 
        while state > 0 {
            if state & 1 == 1 {
                time += jobs[count];
            } 
            state >>= 1; 
            count += 1; 
        }
        time 
    }
    pub fn minimum_time_required(jobs: Vec<i32>, k: i32) -> i32 {
        // dp able 
        let n = jobs.len(); 
        let mut dp = vec![vec![0; (2 as i32).pow(n as u32) as usize]; k as usize];
        // init dp 
        for i in 0..(2 as i32).pow(n as u32) {
            dp[0][i as usize] = Solution::state_to_time(&jobs.as_slice(), i);
        } 
        // traverse 
        for i in 1..k {
            for state in 0..(2 as i32).pow(n as u32) {
                let mut min = std::i32::MAX;
                let mut subset = state; 
                while subset > 0 {
                    min = min.min(dp[i as usize - 1][(state ^ subset) as usize].max(dp[0][subset as usize])); 
                    subset = (subset - 1) & state; 
                }
                dp[i as usize][state as usize] = min; 
            }
        }
        dp[dp.len()-1][dp[0].len()-1]
    }
}

struct Solution {}

fn main() {
    assert_eq!(Solution::minimum_time_required(vec![3,2,3], 3), 3); 
}